lib/scan.ts

```diff
@@ -1,44 +1,76 @@
 import * as fs from 'fs';
-import { fromUrl } from 'hosted-git-info';
 import * as path from 'path';
+
+import { Facts, Options, PluginResponse, ScanResult } from './types';
+import { SignatureResult } from './types';
+import { debug } from './debug';
 import { find } from './find';
+import { fromUrl } from 'hosted-git-info';
+import { getSignature } from './signatures';
 import { getTarget } from './git';
-import { hash } from './hash';
-import { debug } from './debug';
-import {
-  ScanResult,
-  Options,
-  Fingerprint,
-  Facts,
-  PluginResponse,
-} from './types';
+
+const processSignatures = async (
+  signatures: Promise<SignatureResult | null>[],
+  allSignatures: SignatureResult[],
+) => {
+  const signaturesResult = await Promise.all(signatures);
+  const signaturesFiltered = <SignatureResult[]>(
+    signaturesResult.filter((signature) => !!signature)
+  );
+  allSignatures.push(...signaturesFiltered);
+  signatures = [];
```

Comment:
Why the side effects? I'm not sure this abstraction is giving us much, other than making the code a bit more difficult to reason about. I'd just push this logic back into the outer function

Reply:
Note that `processSignatures` is called in two different places. This was abstracted out to avoid code duplication. We could get rid of the side effects, but I think it would probably decrease performance (allocating new arrays vs. modifying "in place").

Reply:
I agree with @robguinness-snyk point, the abstraction is required and helps with performance.